### PR TITLE
a draft of Feature/group messaging (DRAFT)

### PR DESCRIPTION
**Pull request by Qcent**
## Description

Closes Issue 5:
[(https://github.com/https://github.com/Qcent/d6580/issues/5)]

1. Added UserConvos to conversation model as a junction table. 
2. Changed User/Conversation to a Many-to-Many relationship associated through UserConvos.

## Notes on your approach and thought process

Please leave some notes explaining your thought process and your approach to solving this issue.

Because the group chat requires conversations between more than two participants, I changed the User to Conversation relationship from a One-to-Many to a Many-to-Many relationship. I accomplished this by linking them through a UserConvos junction table.

### What additional (non-database) changes are needed to implement this feature (we want this to be high level but thorough)?

1. Update the conversation's API route to find a User by ID, and include associated conversations and messages within. 
2. Front End will need to sort through the "members" of each conversation to replace the ".otherUser" functionality. 
3. The message read status feature could be extended to all "members" by adding a column to the conversation model that would track the last message read by each member.
4.  The sidebar chat preview would have to list multiple usernames and a possible badge for total "members" in the conversation. 
5. There will be a need for a method to invite other "members" to the group chat. 

### If our app were already deployed, what steps would we need to take to move to this new feature without disrupting service for current users?

By writing a simple sequelize script, all existing conversations in database could be updated to use the UserConvos junction table. This method might produce a small amount of downtime while the database updates and users would need to refresh. Another option would be to use a separate model for group chats aside from the existing conversation model, and support both messaging models temporarily, until the database can be switched over seamlessly. 

### Can you be specific about what are the UI features you will implement to reflect the schema change?

1. Chats in the side bar with more than 2 members will need to be displayed differently.  they should be identified as a group chat and display multiple comma separated usernames, possibly truncated with "and n others", where n is the number of usernames not listed.
2. In the active chat header. Group chats should be identified as such. and display names and online status of members in a scrollable or dropdown style list.
3.  Also in the active chat header a button or option from a context menu spawned from the MoreHorizIcon component; That would allow other users to be selected, as is done in the sidebar search, and included as group chat members.
This might involve a modal type component to display the user selections

### How would you update React state to reflect the schema change?

Conversations and messages should be handled in the same manner with changes only needed to be made for the otherUser data field, which I would replace with a members object containing keys of the user ids of all members in the conversation.

### How would you update socket event handlers to reflect the schema change?

I would need socket handlers for users joining and leaving a group chat..

### How would you update route handlers to reflect the schema change?

The conversation route would need to be updated to retrieve a user with all conversations they participate in joined into the data. Instead of the otherUser property being added to the returned data, a members object will be created that will allow access to members data through their user id as a key (to prevent/reduce nested loops). The data stored for each member would include: username, lastMessageRead and online status

### Can you briefly describe the procedure to remove the deprecated models?

Once the database and all users have switched to using the new models all that would be required is to drop the user1 and user2 columns from the conversion table.

### For the migration plan, can you itemize in detail the step by step approach you would adopt to migrate the old data (present in the old database schema) to the new schema without disrupting the service for clients?

1. A sequelize script can be written to find all existing conversations with the user1 and user2 data fields and add each respective user id as a member to the conversion. 
2. Api routes can then be updated to return data differently if there are more than 2 members in conversation. Allowing both old and new schema to function simultaneously.
3. Push the changes to the react front end that handle data from augmented api routes.

### If we have multiple developers involved in this project, how can we split the task between them without introducing merge conflict?

1. Home.js must be edited to refactor state updates and socket handlers.
2. Api routes must be edited.
3. The sidebar and active chat components must be edited.
4. The sequelize database update script needs to be written.

working on these issues separately would not cause any merge conflicts.

### CODE REVIEW

**Review by uimarshall — COMMENTED** [review]
Hi @Qcent,
Thank you for your effort on this PR! so far 👍. Nice work attempting this issue.

- It seems you didn’t add the word “draft” in the title of the PR as was requested In the [issue ](https://github.com/Qcent/d6580/issues/5#:~:text=feature.%20Please%20include-,%22draft%22,-in%20the%20title)description, it asks to have the title of this PR include the word "draft" (since this is not intended to be merged in, but to stay as a draft). Please make sure to add that in.

I've left some additional comments on the code.
Also, here are a few additional questions:

- Can you be specific about what are the UI features you will implement to reflect the schema change?
- How would you update React state to reflect the schema change?
- How would you update socket event handlers to reflect the schema change?
- How would you update route handlers to reflect the schema change?
- Can you briefly describe the procedure to remove the deprecated models?
- For the migration plan, can you itemize in detail the step by step approach you would adopt to migrate the old data (present in the old database schema) to the new schema without disrupting the service for clients?
- If we have multiple developers involved in this project, how can we split the task between them without introducing merge conflict?

Please append the answers to this PR and let us know on Slack when it's ready for a second review 🙂

A small request: please don’t resolve the conversations, but feel free to reply to them.

**Review by temesghentekeste — APPROVED** [review]
Hi @Qcent  :wave: 
Thank you for your work on this PR.

**If we have multiple developers involved in this project, how can we split the task between them without introducing merge conflict?**

- Effective communication is one of the best ways to reduce issues, daily scrum meetings(standup meetings),  using collaboration tools such as Slack, Microsoft Teams, or Discord, and project management tools such as Clickup or Trello could be used.
- To avoid merge conflicts, we can assign issues to different developers and allow them to work in separate git branches for a specific feature. 

**For data migration plan:**

- The data migration plan could take a lot of time and that could cause disruption of service for current users. We should implement an intermediate schema (two versions of the database) and conditionally route the requests to each of them according to the type of conversation.

Consider this PR approved. Please do not merge this PR since it is only a draft. :slightly_smiling_face:

**Pull request closed without merging**
